CatalogReaderExample/NuGet.Protocol.Catalog/CatalogClient.cs

```diff
@@ -0,0 +1,112 @@
+        {
+            // Buffer all of the JSON so we can parse twice. Once to determine the leaf type and once to deserialize
+            // the entire thing to the proper leaf type.
+            _logger.LogDebug($"Downloading {leafUrl} as a byte array.");
+            var jsonBytes = await _httpClient.GetByteArrayAsync(leafUrl);
+            var untypedLeaf = DeserializeBytes<CatalogLeaf>(jsonBytes);
+
+            switch (untypedLeaf.Type)
+            {
+                case CatalogLeafType.PackageDetails:
+                    return DeserializeBytes<PackageDetailsCatalogLeaf>(jsonBytes);
+                case CatalogLeafType.PackageDelete:
+                    return DeserializeBytes<PackageDeleteCatalogLeaf>(jsonBytes);
+                default:
+                    throw new NotSupportedException($"The catalog leaf type '{untypedLeaf.Type}' is not supported.");
+            }
+        }
+
+        private async Task<CatalogLeaf> GetLeafAsync(CatalogLeafType type, string leafUrl)
+        {
+            switch (type)
+            {
+                case CatalogLeafType.PackageDetails:
+                    return await GetPackageDetailsLeafAsync(leafUrl);
+                case CatalogLeafType.PackageDelete:
+                    return await GetPackageDeleteLeafAsync(leafUrl);
+                default:
+                    throw new NotSupportedException($"The catalog leaf type '{type}' is not supported.");
+            }
+        }
+
+        public Task<PackageDeleteCatalogLeaf> GetPackageDeleteLeafAsync(string leafUrl)
+        {
+            return GetAndValidateLeafAsync<PackageDeleteCatalogLeaf>(CatalogLeafType.PackageDelete, leafUrl);
+        }
+
+        public Task<PackageDetailsCatalogLeaf> GetPackageDetailsLeafAsync(string leafUrl)
+        {
+            return GetAndValidateLeafAsync<PackageDetailsCatalogLeaf>(CatalogLeafType.PackageDetails, leafUrl);
+        }
+
+        private async Task<T> GetAndValidateLeafAsync<T>(CatalogLeafType type, string leafUrl) where T : CatalogLeaf
+        {
+            var leaf = await DeserializeUrlAsync<T>(leafUrl);
+
+            if (leaf.Type != type)
+            {
+                throw new ArgumentException(
+                    $"The leaf type found in the document does not match the expected '{type}' type.",
+                    nameof(type));
+            }
+
+            return leaf;
+        }
+
+        private T DeserializeBytes<T>(byte[] jsonBytes)
+        {
+            using (var stream = new MemoryStream(jsonBytes))
+            using (var textReader = new StreamReader(stream))
+            using (var jsonReader = new JsonTextReader(textReader))
+            {
+                return JsonSerializer.Deserialize<T>(jsonReader);
+            }
+        }
+
+        private async Task<T> DeserializeUrlAsync<T>(string documentUrl)
+        {
+            _logger.LogDebug($"Downloading {documentUrl} as a stream.");
```

Comment:
Super nit: can we add a space after the log line? :)

Reply:
Fixed.